### PR TITLE
[DDW-559] Implement delegation deposit from coin selection

### PR DESCRIPTION
**Pull request by tomothespian**
This PR introduces dynamic delegation deposit from coin selection.

---

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1612273279130800)
- [ ] Open Delegation wizard with Trezor and Ledger HW wallets and check if fees are correctly calculated. If you already delegated with that wallet fees are regular and without deposit. Otherwise, deposit is included
- [ ] Delegate wallet for the first time (**with** charged deposit)
- [ ] Delegate wallet one more time (**without** charged deposit)

**_NOTE:_**
_Pay attention to fees. When you want to delegate HW for the very first time, a delegation deposit is also included in the fees. After that deposit is not charged anymore.
Since you already delegated with your HW, to check if a deposit is included (fees > 2ADA) you will need to wipe your devices. When you will wipe devices ensure that you have initial mnemonics on safe (from first device setup) so you can access your funds after this test. Then set a new device (wipe) with different mnemonics. After that, you will get a new fresh wallet / device. Create a new wallet in Daedalus with a wiped device and you will see that balance is 0ADA._

- Please check this one on TestNet and Mainnet

**Reset (wipe) device instructions:**

_Ledger:_
https://support.ledger.com/hc/en-us/articles/360017582434-Reset-to-factory-settings

_Trezor:_
https://wiki.trezor.io/User_manual:Wiping_the_Trezor_device
https://wiki.trezor.io/User_manual:Recovery#Recovery_process

---
## Test Cases

**TREZOR T - Scenario 1 - Validate fees before delegating Trezor T Wallet**
```Gherkin
Given a Trezor T wallet with at least 10 ADA is added
And both Daedalus and the wallet are synced
And the Delegation Center screen on Staking Center is displayed
When I click on "Delegate" button for the wallet
And I follow the instruction on the wizard to delegate the wallet
Then I can see on the last step that the delegation fees includes the deposit (2 ADA)
```

**TREZOR T - Scenario 2 - Validate fees after delegating Trezor T Wallet**
```Gherkin
Given I am on the confirmation dialog for delegating a Trezor T wallet
And the fees displayed include the deposit for delegation
When I plug, unlock and export the public key on the device
And confirm the transaction on the dialog
Then the wallet is displayed as delegated to the new stake pool in the epoch number = current epoch + 2
And the amount of the transaction on the "Transaction" tab matches the displayed on the confirmation dialog
```

**TREZOR T - Scenario 3 - Validate fees before redelegating Trezor T Wallet**
```Gherkin
Given a Trezor T wallet with at least 10 ADA is added
And both Daedalus and the wallet are synced
And the wallet is already delegated
And the Delegation Center screen on Staking Center is displayed
When I click on "Delegate" button for the wallet
And I follow the instruction on the wizard to delegate the wallet
Then I can see on the last step that the delegation fees don't include a deposit
```

**LEDGER NANO S - Scenario 1 - Validate fees before delegating Ledger Nano S Wallet**
```Gherkin
Given a Ledger Nano S wallet with at least 10 ADA is added
And both Daedalus and the wallet are synced
And the Delegation Center screen on Staking Center is displayed
When I click on "Delegate" button for the wallet
And I follow the instruction on the wizard to delegate the wallet
Then I can see on the last step that the delegation fees includes the deposit (2 ADA)
```

**LEDGER NANO S - Scenario 2 - Validate fees after delegating Ledger Nano S Wallet**
```Gherkin
Given I am on the confirmation dialog for delegating a Ledger Nano S wallet
And the fees displayed include the deposit for delegation
When I plug, unlock and export the public key on the device
And confirm the transaction on the dialog
Then the wallet is displayed as delegated to the new stake pool in the epoch number = current epoch + 2
And the amount of the transaction on the "Transaction" tab matches the displayed on the confirmation dialog
```

**LEDGER NANO S - Scenario 3 - Validate fees before redelegating Ledger Nano S Wallet**
```Gherkin
Given a Ledger Nano S wallet with at least 10 ADA is added
And both Daedalus and the wallet are synced
And the wallet is already delegated
And the Delegation Center screen on Staking Center is displayed
When I click on "Delegate" button for the wallet
And I follow the instruction on the wizard to delegate the wallet
Then I can see on the last step that the delegation fees don't include a deposit
````

**LEDGER NANO X - Scenario 1 - Validate fees before delegating Ledger Nano X Wallet**
```Gherkin
Given a Ledger Nano X wallet with at least 10 ADA is added
And both Daedalus and the wallet are synced
And the Delegation Center screen on Staking Center is displayed
When I click on "Delegate" button for the wallet
And I follow the instruction on the wizard to delegate the wallet
Then I can see on the last step that the delegation fees includes the deposit (2 ADA)
```

**LEDGER NANO X - Scenario 2 - Validate fees after delegating Ledger Nano X Wallet**
```Gherkin
Given I am on the confirmation dialog for delegating a Ledger Nano X wallet
And the fees displayed include the deposit for delegation
When I plug, unlock and export the public key on the device
And confirm the transaction on the dialog
Then the wallet is displayed as delegated to the new stake pool in the epoch number = current epoch + 2
And the amount of the transaction on the "Transaction" tab matches the displayed on the confirmation dialog
```

**LEDGER NANO X - Scenario 3 - Validate fees before redelegating Ledger Nano X Wallet**
```Gherkin
Given a Ledger Nano X wallet with at least 10 ADA is added
And both Daedalus and the wallet are synced
And the wallet is already delegated
And the Delegation Center screen on Staking Center is displayed
When I click on "Delegate" button for the wallet
And I follow the instruction on the wizard to delegate the wallet
Then I can see on the last step that the delegation fees don't include a deposit
```
---
### Basics

- [ ] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes
- [ ] PR has default-sized Daedalus window screenshots or animated GIFs of important UI changes:
  - [ ] In English
  - [ ] In Japanese
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

### CODE REVIEW

File: source/renderer/app/api/api.js [hunk 1]
```diff
@@ -964,16 +966,26 @@ export default class AdaApi {
         LOVELACES_PER_ADA
       );
 
+      let delegationDepositSum = 0;
```
Comment (nikolaglumac) [inline source/renderer/app/api/api.js:969]: @tomothespian please change this to "totalDeposits" and push it above the line where you calculate `transactionFee`: https://github.com/input-output-hk/daedalus/blob/728ec6e0978bb4fd34e21d61fb29abc925768ba5/source/renderer/app/api/api.js#L965-L967

Then you will have a nicer way to calculate the fee:
```
const fee = new BigNumber(totalInputs - totalOutputs - totalDeposits).dividedBy(
    LOVELACES_PER_ADA
);
```

File: source/renderer/app/api/api.js [hunk 2]
```diff
@@ -964,16 +966,26 @@ export default class AdaApi {
...
+      if (response.deposits) {
```
Comment (nikolaglumac) [inline source/renderer/app/api/api.js:970]: @tomothespian please use a nicer way to sum up all deposits. Something like:

```
const totalDeposits = map(response.deposits || [], (deposit) => {...
```

File: source/renderer/app/api/api.js [hunk 3]
```diff
@@ -964,16 +966,26 @@ export default class AdaApi {
...
+        map(response.deposits, (deposit) => {
+          delegationDepositSum += deposit.quantity;
+        });
+      }
+      const formattedDelegationDeposit = formattedLovelaceToAmount(
+        delegationDepositSum
+      );
+      const delegationDeposit = new BigNumber(formattedDelegationDeposit);
+
+      // On first wallet delegation fee also includes a deposit
```
Comment (nikolaglumac) [inline source/renderer/app/api/api.js:980]: @tomothespian we don't need this comment anymore...

**Review by nikolaglumac — CHANGES_REQUESTED** [review]
@tomothespian please implement requested changes 🙏

**Comment by gnpf** [comment]
Works as expected on Ledger Nano S 💯

**Comment by mioriohk** [comment]
LGTM on Ledger Nano X. Good job @tomothespian 👍 . Tested on 16219

**Review by ManusMcCole — APPROVED** [review]
All good for Trezor Model T on 16219 👍

**Comment by tomothespian** [comment]
@nikolaglumac Code Improved!

@gnpf @ManusMcCole @mioriohk Can you test one more time just in case

**Review by nikolaglumac — APPROVED** [review]
Great work @tomothespian 💯

File: source/renderer/app/api/api.js [hunk 4]
```diff
@@ -960,26 +959,21 @@ export default class AdaApi {
...
-      if (delegation && delegation.delegationAction) {
-        const delegationDeposit = new BigNumber(DELEGATION_DEPOSIT);
-        const isDepositIncluded = fee.gt(delegationDeposit);
-        transactionFee = isDepositIncluded ? fee.minus(delegationDeposit) : fee;
-      } else {
-        transactionFee = fee;
-      }
+      const totalDeposits = sumBy(response.deposits, 'quantity');
+      const feeWithDeposits = new BigNumber(
+        totalInputs - totalOutputs
```
Comment (nikolaglumac) [inline source/renderer/app/api/api.js:965]: @tomothespian please use BigNumber to deduct outputs from inputs.

File: source/renderer/app/api/api.js [hunk 5]
```diff
@@ -960,26 +959,21 @@ export default class AdaApi {
...
+      ).dividedBy(LOVELACES_PER_ADA);
+      const fee = new BigNumber(
+        totalInputs - totalOutputs - totalDeposits
```
Comment (nikolaglumac) [inline source/renderer/app/api/api.js:968]: @tomothespian please use BigNumber to deduct deposits and outputs from inputs.

**Review by nikolaglumac — CHANGES_REQUESTED** [review]

**Review by mioriohk — APPROVED** [review]
LGTM on Ledger Nano X. Good job @tomothespian 👍 . Tested on 16268

**Review by gnpf — APPROVED** [review]

**Review by ManusMcCole — APPROVED** [review]

**Review by nikolaglumac — APPROVED** [review]
Great work @tomothespian 💯

**Pull request merged**